Assignment statements

zygoloid opened the pull request:
Assignment is permitted only as a complete statement, not as a subexpression.
Assignment and compound assignment syntax follow C++ in all other respects.
Pre-increment is provided. Post-increment is not. Uses of all of these operators
are translated into calls to interface members.

zygoloid committed:
Filling out template with PR 2511

zygoloid committed:
Proposal.

zygoloid committed:
Add alternatives considered to design.

zygoloid committed:
Remove period from heading

zygoloid committed:
Fix code formatting

zygoloid committed:
Fix typo in rewrite rule.

zygoloid committed:
Make indentation consistent.

zygoloid committed:
Remove period for consistency

carsonRadtke commented on docs/design/assignment.md:
````diff
@@ -0,0 +1,329 @@
+```
+
+For each binary [arithmetic](expressions/arithmetic.md) or
+[bitwise](expressions/bitwise.md) operator `$`, a corresponding compound
````
I like the usage of "`$`" and "`$=`" below, but I think "`$`" is a distraction here because there is no accompanying "`$=`".

zygoloid replied:
Good point, added a `$=` here.

zygoloid commented on docs/design/assignment.md:
````diff
@@ -0,0 +1,329 @@
+```
+
+For each binary [arithmetic](expressions/arithmetic.md) or
+[bitwise](expressions/bitwise.md) operator `$`, a corresponding compound
+assignment is provided that performs the operation in-place:
````
```suggestion
[bitwise](expressions/bitwise.md) operator `$`, a corresponding compound
assignment `$=` is provided that performs the operation in-place:
```

zygoloid committed:
Add use of `$=` to match introduction of `$`.

chandlerc commented on docs/design/README.md:
```diff
@@ -1223,25 +1222,6 @@ fn Foo() {
 > -   Proposal
 >     [#162: Basic Syntax](https://github.com/carbon-language/carbon-lang/pull/162)
 
-### Assignment statements
```
Any reason to not keep at least a brief overview here?

zygoloid replied:
We have a very brief overview on line 253-260. I didn't feel like this section pulled its weight compared to that very brief overview and the detailed description in the separate page.

chandlerc replied:
FWIW, I still would find it helpful, although its not a big deal. Maybe @josh11b has a thought one way or the other?

Assuming we don't keep it, I think then the section above or up one level should be adjusted to at least have a list of the different kinds of statements and links to their dedicated page. Otherwise, folks dropping in this section from a link to that section can't really navigate to all the relevant parts of the design.

chandlerc replied:
@josh11b to maybe provide guidance on the right amount of content and structure here.

FWIW, the only thing I'm really looking for here is some reasonably discoverable link to the dedicated document. Totally happy with however much or little detail accompanies it here and whatever structure works best to incorporate it. =]

josh11b replied:
I do think there should be something inside the previous section about assignment statements if this section is going to be removed, like around the sentence starting with "Expressions."

zygoloid replied:
Thanks, that makes sense to me. Done.

chandlerc commented on docs/design/assignment.md:
````diff
@@ -0,0 +1,329 @@
+}
+```
+
+## Built-in types
````
What are built-in types? I would expect that we'd use all of same interface stuff via the prelude even for integers and such and just have the implementations of the interface functions be built-in?

zygoloid replied:
This is the same style of presentation we use elsewhere in the design.  Eg: https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/expressions/arithmetic.md#built-in-types

Per the extensibility section below: "Implementations of these interfaces are provided for built-in types as necessary to give the semantics described above."

zygoloid replied:
I've reworked this section a bit to be clearer about what types are in scope here.

chandlerc commented on docs/design/assignment.md:
````diff
@@ -0,0 +1,329 @@
+
+```
+// Simple `=`.
+interface AssignFrom(U:! type) {
````
I find the difference of `AssignFrom` vs `AssignWith` a bit odd .. is there a wording that would work for both?

zygoloid replied:
I don't think we can easily get away from `=` and `$=` being somewhat different here, but perhaps reordering the words would help:

- `Assign` and `AssignFrom` unchanged
- `AssignAdd` and `AssignAddWith`, instead of `AddAssign` and `AddAssignWith`

Using `AssignAddWith` instead of `AddAssignWith` seems more harmonious with `AddWith` I think it makes the `Assign` / `AssignFrom` pair look less strange by making it more different. `AssignAddWith` seems a bit better than `AddAssignWith` to me, `AssignAdd` seems a little worse than `AddAssign`; on balance I think I like these reordered names more.

We could go further and replace `Assign` with `InPlace` for the compound forms:

- `Assign` and `AssignFrom`
- `InPlaceAdd` and `InPlaceAddWith`

I think I'd be happy with that too, but I think we lose something by having `=` in the operator symbol but not having `Assign` in the name.

Does one of those options work better for you?

Pixep replied:
Not fully satisfying, but maybe `AssignGiven` and `AddAssignGiven`

zygoloid replied:
After considering a bunch of options, I've settled on `Assign` / `AssignWith` / `AssignOp` / `AssignOpWith` being the most consistent set of names. I'm not entirely happy with these names, but I'm also not sure we can do better.

zygoloid replied:
After much discussion, reversed `AssignOp` back to `OpAssign`.

chandlerc commented on proposals/p2511.md:
```diff
@@ -0,0 +1,190 @@
+to exist, and they may more directly convey the intended semantics of counting
+than a `+=` operation would.
+
+### Define `$=` in terms of `$`
```
FWIW, I suspect this will be a pain point for folks. I'm guessing there will be a lot of types where the potential costs are either immaterial or actually don't exist and the conceptual model will be substantially cleaner the other way.

Another thing to think about is how a generic function needs to be constrained. It would seem much more obvious to teach that you require `Add & Mul & Assign` and get `+`, `*`, `=`, `+=`, `*=`. The other way around seems likely to be much less obvious.

I wonder if this should even override the concern around performance.

jonmeow replied:
I may be confused, but doesn't the "Default" section of the design address this? (though, this alternative probably should mention the default interfaces)

zygoloid replied:
@jonmeow My understanding is that @chandlerc would like to get `$=` defined in terms of `$`, whereas this proposal previously did the opposite: it defined `$` in terms of `$=`.

@chandlerc I think you've made a good point here, and in fact I'm not sure the performance argument actually holds up. I can easily think of situations where for performance you wouldn't want either operation defined in terms of the other one. Defining `$=` in terms of `$` and `=` seems to be strictly better if we set that aside. I've made that change.

chandlerc commented on proposals/p2511.md:
```diff
@@ -0,0 +1,190 @@
+-   Destroy the left-hand operand.
+-   Move the value created earlier into the left-hand operand.
+
+However, this removes some flexibility and could harm performance, for example
```
Maybe worth mentioning that it'd also much more sharply deviate from C++ where types are currently designed with this flexibility in mind?

zygoloid replied:
Done.

chandlerc reviewed (commented):
Some quick initial thoughts...

jonmeow commented on proposals/p2511.md:
```diff
@@ -0,0 +1,190 @@
+
+## Alternatives considered
+
+### Allow assignment as a subexpression
```
I will note particular concerns with this for increment and decrement. e.g., [absl's flat_hash_map erase](https://github.com/abseil/abseil-cpp/blob/master/absl/container/flat_hash_map.h#L229-L233) relies on postincrement as a subexpression. This could make particular challenges for migrating code, as either adding additional locals (since `map.erase(it); ++it;` is potentially different) or creating new APIs becomes a prerequisite for migration, likely reducing automated migration.

But it's not clear to me that you're looking for this kind of concern at this time, so I'll mostly suggest that if you're still considering this, "alternatives considered" indicates that it has been ruled out; "open questions" might be more appropriate? I think the difference here is whether the plan is "look more" versus "wait and see if there are too many complaints"; if it's the latter, I would agree with "alternatives considered", but would also want more consideration for how we migrate existing C++ code on the assumption that this isn't reconsidered.

jonmeow replied:
Thought: for migration, maybe we just transform C++ `it++` into something like `[&]() -> auto { var x: auto = it; ++it; return x; }()` (something like that can become a general solution for subexpressions, although maybe not ideal for readers)

zygoloid replied:
Thanks, I've added some brief discussion here giving the rationale for rejecting this alternative, which was previously missing, and also describing possibilities for getting the same effect (eg, for migration) and some potential limitations when doing so.

chandlerc replied:
What do you think about having a distinct alternative for increment and decrement? It can I think be much more brief, but I feel like there are some unique aspects to those, both in terms of the impact and potentially more limited pitfalls. And so it might be clearer to treat those separately.

Minor point either way though -- there is a fundamental tradeoff here in that this will create some unknown amount of migration friction, and it'll be experience with tooling and actual migrations that teach us whether we end up want to revisit parts of this or not.

jonmeow replied:
Just to be explicit, I'm fine with the resolution here, although apprehensive about migration costs.

jonmeow commented on docs/design/assignment.md:
```diff
@@ -0,0 +1,329 @@
+
+## Syntax
+
+The operands of these operators can be any [expression](expressions/README.md).
```
Are there any parsing concerns, e.g. how is `a*=b` interpreted? Is that a postfix `*` (as in `i32*`) or `*=`?

zygoloid replied:
That ends up being a lexical consideration, not a parsing one: `*=` is a single token, regardless of whitespace before the `*.` Per [the current lexing design](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design/lexical_conventions), as decided in #601, we always form the longest lexical element we can, so there's no possibility of forming a `*` token when the next two characters are `*=`.

This does mean that `var v: i32*=nullptr;` is invalid. I don't think that's a major problem, given that we encourage whitespace around binary operator tokens like `=` anyway.

jonmeow commented on docs/design/assignment.md:
````diff
@@ -0,0 +1,329 @@
+```
+
+This default is not sufficient to provide implicit conversions on the left-hand
+side of an operator. If such conversions are desired, as separate implementation
+is necessary:
````
```suggestion
side of an operator. If such conversions are desired, a separate implementation
is necessary:
```

jonmeow commented on proposals/p2511.md:
```diff
@@ -0,0 +1,190 @@
+We could restrict this to only apply to simple assignment, not cases like
+`a += b -= c *= 2;`
+
+In leads issue [#451](https://github.com/carbon-language/carbon-lang/issues/451)
+it was decided that we would initially not support chained assignment.
```
Should the rationale be provided here so that leads issues aren't a canonical reference for decisions?

Perhaps more importantly, what _was_ the rationale? What trade-offs did the leads consider? On a skim, [I see a decision](https://github.com/carbon-language/carbon-lang/issues/451#issuecomment-837483823), but without further context.

zygoloid replied:
Added some words, but honestly I'm not sure we had much rationale here other than that we didn't see sufficient motivation to add this feature compared to the cost.

jonmeow replied:
A lack of justification is still important to document. :)

zygoloid committed:
Fix typo.

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>

zygoloid committed:
Address review comments. Rename interfaces.

zygoloid committed:
Fix constraint to put `OpWith` in scope before we use it.

chandlerc commented on proposals/p2511.md:
```diff
@@ -0,0 +1,321 @@
+// ✅ OK, per pending proposal #2006.
+a = 1;
+// 🤷 Might result in an error or warning;
+// `m` is in an unformed state.
```
```suggestion
// `b` is in an unformed state.
```

chandlerc reviewed (commented):
Just noting a few things here orthogonal to the naming discussion...

jonmeow commented on proposals/p2511.md:
```diff
@@ -132,16 +134,49 @@ rules, but those rules are not part of this proposal.
+Carbon's eventual lambda syntax.
+
+Workarounds that are not assignment statements would likely be treated as
+capturing `x` for static analysis puurposes, not as definitely initializing `x`,
```
```suggestion
capturing `x` for static analysis purposes, not as definitely initializing `x`,
```

jonmeow reviewed (commented):
LGTM overall.

zygoloid committed:
Apply suggestions from code review

Co-authored-by: Jon Ross-Perkins <jperkins@google.com>
Co-authored-by: Chandler Carruth <chandlerc@gmail.com>

zygoloid committed:
AssignOp -> OpAssign

zygoloid committed:
Fix typos.

zygoloid committed:
Trying bold instead of italics.

zygoloid committed:
Bold+italic.

zygoloid committed:
Fix

zygoloid committed:
We have no lvalue conditionals; remove suggestion that we do.

chandlerc commented on docs/design/assignment.md:
```diff
@@ -0,0 +1,332 @@
+See leads issue
+[#686: Operation order in struct/class assignment/initialization](https://github.com/carbon-language/carbon-lang/issues/686)
+
+## Extensibility
+
+Assignment operators can be provided for user-defined types by implementing the
+following families of interfaces. Implementations of these interfaces are
+provided for built-in types as necessary to give the semantics described above.
+
```
I think it's worth documenting the design model we discussed today here -- here is my very rough sketch. Feel free to rewrite though, not sure if this is the best way of expressing it.

```suggestion
## Extensibility

Assignment operators can be provided for user-defined types by implementing the
following families of interfaces. Implementations of these interfaces are
provided for built-in types as necessary to give the semantics described above.

The *compound* assignment operators are expected to fundamentally be syntactic sugar in all cases, and default implementations of them will be synthesized from relevant assignment and operation implementations where available. There is no support for extending types to have compound assignment operators such as `a += b` that do not correspond to the relevant decomposed operators as in `a = a + b`.

In contrast, increment and decrement which are independently extensible. While for most numeric types, increment for example will map to `a = a + 1`, this is not universal or something Carbon defaults. Floating point types are one example where this doesn't hold true, as are types such as non-random-access iterators.
```

zygoloid replied:
Added something along these lines.

chandlerc commented on proposals/p2511.md:
```diff
@@ -0,0 +1,330 @@
+Overall, the proposed set of names seem like the best choice, despite some of
+the names not reading completely naturally. The English readability concern is
+probably not much worse than for `LeftShiftWith`, where we already decided that
+consistency was more important.
```
Again, trying to incorporate some of our discussion, maybe an added alternative (will also need a link in the design):

```suggestion
consistency was more important.

### Default increment and decrement when we can add or subtract a literal `1`

We could provide a default blanket `impl` for incrementing and decrementing whenever we can rewrite it to `a += 1` or `a -= 1` respectively.

Our decision here will result in a trade off between the expressivity provided by a narrow set of generic constraints and the expressivity of the type interface extension. With this alternative, it might be easier to constrain a generic such that both `a += x` and `++a` are valid. However, it would no longer be possible to express a type which cannot be incremented but which can have `1` added to it.

Ultimately, the benefit to generic code is minimal as the cost of an additional constraint to use `++a` or simply writing `a += 1` if preferable seems easily paid. On the other hand, we already have a goal of not providing the increment operator for floating point types due to the potential confusion over adding `1` vs. stepping to the next distinct floating point value. As a result, we keep increment and decrement independent extensible operations, much like we keep subtraction independent from addition combined with negation.
```

zygoloid replied:
Added something along these lines.

chandlerc reviewed (commented):
Definitely converging. Mostly suggesting some added information to cover a bit of discussion that we had around extensibility of increment and decrement to try and more fully document the principles that underpin the design and the rationale for those.

I also think we need at least some fixes to the top level overview.

zygoloid committed:
Add example of increment in the tour of the language. Add assignment
statements to the list of things that can appear in a block.

zygoloid committed:
Add writeup of intended compound assignment and increment semantics.

Add alternative considered for providing increment automatically
whenever we can add 1.

chandlerc reviewed (approved):
LGTM! I spotted a tiny wording nit just as I was skimming it, but seems trivial to fix forward if the suggestion isn't quite right. =D

chandlerc commented on docs/design/assignment.md:
```diff
@@ -0,0 +1,357 @@
+    resources that might not be available, such as a context object or an
+    allocator.
+
+This syntactic sugar is provided by a [default implementation](#defaults) of
+`$=` in terms of `$` and `=`.
```
Actually hitting the button to submit my wording tweak....

```suggestion
The syntactic sugar implementation is provided by a [default implementation](#defaults) of
`$=` in terms of `$` and `=`.
```

zygoloid replied:
```suggestion
The syntactic sugar is implemented by a [default implementation](#defaults)
of `$=` in terms of `$` and `=`.
```

chandlerc replied:
LG

chandlerc reviewed (approved):
(sorry for not including this tweak...)

zygoloid committed:
Minor wording tweak.

zygoloid commented on docs/design/assignment.md:
```diff
@@ -0,0 +1,357 @@
+    resources that might not be available, such as a context object or an
+    allocator.
+
+The syntactic sugar is implemented by a [default implementation](#defaults)
+of `$=` in terms of `$` and `=`.
```
```suggestion
The syntactic sugar is implemented by a [default implementation](#defaults) of
`$=` in terms of `$` and `=`.
```

zygoloid committed:
Appease prettier.
